File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/binarysearch/Problem2343.java
package com.keunwon.algorithm.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Problem2343 {
    private static int N, M;
    private static int[] times;

    private static int left, right;

    public static void main(String[] args) throws IOException {
        input();
        binarySearch();
    }

    private static void binarySearch() {
        while (left <= right) {
            final var mid = (left + right) / 2;

            if (check(mid)) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        System.out.println(left);
    }

    private static boolean check(int size) {
        var count = 0;
        var sum = 0;

        for (int i = 0; i < N; i++) {
            final var num = times[i];

            if (sum + num <= size) {
                sum += num;
                continue;
            }

            count++;
            sum = num;
        }
        return M <= count;
    }

    private static void input() throws IOException {
        final var br = new BufferedReader(new InputStreamReader(System.in));

        try (br) {
            var st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            M = Integer.parseInt(st.nextToken());
            times = new int[N];

            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < N; i++) {
                final var num = Integer.parseInt(st.nextToken());
                times[i] = num;

                left = Math.max(left, num);
                right += num;
            }
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2343.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: 기타 레슨
 * Level: 실버-1
 **/
class Problem2343 {
    fun solution(m: Int, arr: IntArray): Int {
        var left = arr.maxOf { it }
        var right = arr.sum()

        while (left <= right) {
            val mid = (left + right) / 2
            val count = getCount(mid, arr)

            if (m >= count) right = mid - 1
            else left = mid + 1
        }
        return left
    }

    private fun getCount(limit: Int, arr: IntArray): Int {
        var count = 0
        var sum = 0

        for (num in arr) {
            if (limit < sum + num) {
                sum = 0
                count++
            }
            sum += num
        }
        return if (sum == 0) count else count + 1
    }
}

fun main() {
    val (n, m) = readLine()!!.split(" ").map { it.toInt() }
    val arr = readLine()!!.split(" ").map { it.toInt() }.toIntArray()

    Problem2343().solution(m, arr)
}
